File: src/types/atomic-events.ts
/**
 * proposal5.md 구현 - "Dual Channel" 이벤트 시스템의 Atomic Task Events
 * 
 * 고빈도 원자적 작업 이벤트를 위한 TypeScript 타입 정의
 * Rust 백엔드의 AtomicTaskEvent enum과 동기화
 */

/** UUID 타입 정의 */
export type UUID = string;

/**
 * 원자적 작업 이벤트 타입 - 개별 작업의 생명주기 추적
 */
export interface AtomicTaskEvent {
  /** 작업 식별자 */
  task_id: UUID;
  /** 이벤트 타임스탬프 (ISO 8601) */
  timestamp: string;
  /** 이벤트 종류 */
  event_type: AtomicEventType;
}

/**
 * 원자적 이벤트 타입 열거형
 */
export type AtomicEventType = 
  | { type: 'TaskStarted' }
  | { type: 'TaskCompleted'; data: TaskCompletedData }
  | { type: 'TaskFailed'; data: TaskFailedData }
  | { type: 'TaskRetrying'; data: TaskRetryingData };

/**
 * 작업 완료 이벤트 데이터
 */
export interface TaskCompletedData {
  /** 처리 시간 (밀리초) */
  duration_ms: number;
  /** 추출된 데이터 크기 */
  data_size: number;
}

/**
 * 작업 실패 이벤트 데이터
 */
export interface TaskFailedData {
  /** 오류 메시지 */
  error: string;
  /** 재시도 가능 여부 */
  retryable: boolean;
}

/**
 * 작업 재시도 이벤트 데이터
 */
export interface TaskRetryingData {
  /** 현재 재시도 횟수 */
  attempt: number;
  /** 이전 오류 메시지 */
  previous_error: string;
  /** 다음 재시도까지 대기 시간 (밀리초) */
  delay_ms: number;
}

/**
 * 원자적 이벤트 통계
 */
export interface AtomicEventStats {
  /** 총 이벤트 수 */
  total_events: number;
  /** 활성 작업 수 */
  active_tasks: number;
  /** 완료된 작업 수 */
  completed_tasks: number;
  /** 실패한 작업 수 */
  failed_tasks: number;
  /** 평균 처리 시간 (밀리초) */
  avg_duration_ms: number;
}

File: src-tauri/src/domain/atomic_events.rs
//! Atomic task events for real-time UI visualization
//!
//! This module defines lightweight, high-frequency events that are emitted
//! immediately when individual tasks change state, enabling real-time
//! animation and detailed progress tracking in the UI.

use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Unique identifier for individual tasks
pub type TaskId = Uuid;

/// Lightweight event for immediate task state changes
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AtomicTaskEvent {
    TaskStarted {
        task_id: TaskId,
        task_type: String,
        timestamp: chrono::DateTime<chrono::Utc>,
    },
    TaskCompleted {
        task_id: TaskId,
        task_type: String,
        duration_ms: u64,
        timestamp: chrono::DateTime<chrono::Utc>,
    },
    TaskFailed {
        task_id: TaskId,
        task_type: String,
        error_message: String,
        retry_count: u32,
        timestamp: chrono::DateTime<chrono::Utc>,
    },
    TaskRetrying {
        task_id: TaskId,
        task_type: String,
        retry_count: u32,
        delay_ms: u64,
        timestamp: chrono::DateTime<chrono::Utc>,
    },
}

impl AtomicTaskEvent {
    /// Get the event name for Tauri emission
    pub fn event_name() -> &'static str {
        "atomic-task-update"
    }

    /// Get the task ID from any variant
    pub fn task_id(&self) -> TaskId {
        match self {
            AtomicTaskEvent::TaskStarted { task_id, .. } => *task_id,
            AtomicTaskEvent::TaskCompleted { task_id, .. } => *task_id,
            AtomicTaskEvent::TaskFailed { task_id, .. } => *task_id,
            AtomicTaskEvent::TaskRetrying { task_id, .. } => *task_id,
        }
    }

    /// Get the task type from any variant
    pub fn task_type(&self) -> &str {
        match self {
            AtomicTaskEvent::TaskStarted { task_type, .. } => task_type,
            AtomicTaskEvent::TaskCompleted { task_type, .. } => task_type,
            AtomicTaskEvent::TaskFailed { task_type, .. } => task_type,
            AtomicTaskEvent::TaskRetrying { task_type, .. } => task_type,
        }
    }

    /// Create a TaskStarted event
    pub fn started(task_id: TaskId, task_type: String) -> Self {
        Self::TaskStarted {
            task_id,
            task_type,
            timestamp: chrono::Utc::now(),
        }
    }

    /// Create a TaskCompleted event
    pub fn completed(task_id: TaskId, task_type: String, duration_ms: u64) -> Self {
        Self::TaskCompleted {
            task_id,
            task_type,
            duration_ms,
            timestamp: chrono::Utc::now(),
        }
    }

    /// Create a TaskFailed event
    pub fn failed(
        task_id: TaskId,
        task_type: String,
        error_message: String,
        retry_count: u32,
    ) -> Self {
        Self::TaskFailed {
            task_id,
            task_type,
            error_message,
            retry_count,
            timestamp: chrono::Utc::now(),
        }
    }

    /// Create a TaskRetrying event
    pub fn retrying(task_id: TaskId, task_type: String, retry_count: u32, delay_ms: u64) -> Self {
        Self::TaskRetrying {
            task_id,
            task_type,
            retry_count,
            delay_ms,
            timestamp: chrono::Utc::now(),
        }
    }
}

/// Statistics for atomic task events (for performance monitoring)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AtomicEventStats {
    pub events_emitted: u64,
    pub events_per_second: f64,
    pub last_emission_time: chrono::DateTime<chrono::Utc>,
    pub event_type_counts: std::collections::HashMap<String, u64>,
}

impl Default for AtomicEventStats {
    fn default() -> Self {
        Self {
            events_emitted: 0,
            events_per_second: 0.0,
            last_emission_time: chrono::Utc::now(),
            event_type_counts: std::collections::HashMap::new(),
        }
    }
}

impl AtomicEventStats {
    /// Record a new event emission
    pub fn record_emission(&mut self, event: &AtomicTaskEvent) {
        let now = chrono::Utc::now();
        let time_diff = (now - self.last_emission_time).num_milliseconds() as f64 / 1000.0;

        self.events_emitted += 1;
        self.last_emission_time = now;

        // Update events per second (exponential moving average)
        if time_diff > 0.0 {
            let current_rate = 1.0 / time_diff;
            self.events_per_second = 0.1 * current_rate + 0.9 * self.events_per_second;
        }

        // Update event type counts
        let event_type = match event {
            AtomicTaskEvent::TaskStarted { .. } => "started",
            AtomicTaskEvent::TaskCompleted { .. } => "completed",
            AtomicTaskEvent::TaskFailed { .. } => "failed",
            AtomicTaskEvent::TaskRetrying { .. } => "retrying",
        };

        *self
            .event_type_counts
            .entry(event_type.to_string())
            .or_insert(0) += 1;
    }
}
